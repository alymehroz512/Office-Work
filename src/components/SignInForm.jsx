import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { loginUser } from "../features/auth/authSlice";
import ParticleNetwork from "./ParticleNetwork";

const SignInForm = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { loading, error, token } = useSelector((state) => state.auth);

  const [role, setRole] = useState("admin");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [showPassword, setShowPassword] = useState(false);
  const [formError, setFormError] = useState("");

  // Redirect if already logged in (now using cookie-based token)
  useEffect(() => {
    if (token) {
      navigate("/");
    }
  }, [token, navigate]);

  const handleSubmit = async (e) => {
    e.preventDefault();
    setFormError("");

    if (!role) {
      setFormError("Please select a role.");
      return;
    }

    if (!email.trim() || !password.trim()) {
      setFormError("Please enter both email and password.");
      return;
    }

    try {
      await dispatch(loginUser({ role, email, password })).unwrap();
      navigate("/");
    } catch (err) {
      console.error("Login failed:", err);
    }
  };

  const getErrorMessage = () => {
    if (!error) return null;
    if (error.includes("Failed to fetch")) {
      return "Failed to fetch";
    }
    return error;
  };

  return (
    <div className="container-fluid SignInFormBackground">
      <ParticleNetwork />
      <div className="row justify-content-center">
        <div className="col-md-5 p-2">
          <div className="card p-5 shadow SignInFormCard">
            <div className="lock-icon-wrapper text-center mb-3">
              <i className="fa-solid fa-lock fs-4" />
            </div>
            <h2 className="signin-heading text-center mb-4">Sign In</h2>

            {/* Role selection */}
            <div className="mb-4 text-center">
              <div className="btn-group" role="group" aria-label="Role selection">
                <button
                  type="button"
                  className={`btn btn-outline-primary d-flex align-items-center gap-2 ${
                    role === "admin" ? "active" : ""
                  }`}
                  onClick={() => setRole("admin")}
                >
                  <i className="fa-solid fa-user-shield" /> Admin
                </button>
                <button
                  type="button"
                  className={`btn btn-outline-primary d-flex align-items-center gap-2 ${
                    role === "manager" ? "active" : ""
                  }`}
                  onClick={() => setRole("manager")}
                >
                  <i className="fa-solid fa-user-tie" /> Manager
                </button>
              </div>
            </div>

            <form onSubmit={handleSubmit} className="form-sm" data-testid="sign-in-form">
              {/* Email */}
              <div className="form-floating mb-3">
                <input
                  type="email"
                  className="form-control"
                  id="floatingEmail"
                  placeholder="name@example.com"
                  value={email}
                  onChange={(e) => setEmail(e.target.value)}
                  data-testid="email-input"
                />
                <label htmlFor="floatingEmail">Email</label>
              </div>

              {/* Password */}
              <div className="form-floating mb-3 position-relative">
                <input
                  type={showPassword ? "text" : "password"}
                  className="form-control"
                  id="floatingPassword"
                  placeholder="Password"
                  value={password}
                  onChange={(e) => setPassword(e.target.value)}
                  data-testid="password-input"
                />
                <label htmlFor="floatingPassword">Password</label>
                <span
                  className="position-absolute top-50 end-0 translate-middle-y me-3"
                  style={{ cursor: "pointer" }}
                  onClick={() => setShowPassword(!showPassword)}
                  data-testid="toggle-password"
                >
                  {showPassword ? (
                    <i className="fa-solid fa-eye-slash" />
                  ) : (
                    <i className="fa-solid fa-eye" />
                  )}
                </span>
              </div>

              {/* Submit Button */}
              <button
                type="submit"
                className="btn btn-primary w-100 d-flex justify-content-center align-items-center gap-2"
                disabled={loading}
                data-testid="sign-in-button"
              >
                {loading ? (
                  <>
                    <i className="fa-solid fa-spinner fa-spin" />
                    Signing In...
                  </>
                ) : (
                  "Sign In"
                )}
              </button>

              {/* Error messages */}
              {formError && (
                <div className="text-danger text-center mt-3" data-testid="form-error">
                  {formError}
                </div>
              )}
              {error && (
                <div className="text-danger text-center mt-2" data-testid="server-error">
                  {getErrorMessage()}
                </div>
              )}
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SignInForm;
