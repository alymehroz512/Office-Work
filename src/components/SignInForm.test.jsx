import React from "react";
import { describe, expect, it, beforeEach, afterEach } from "@jest/globals";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import authReducer from "../features/auth/authSlice";
import SignInForm from "../components/SignInForm";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";

// Mock ParticleNetwork component
jest.mock('./ParticleNetwork', () => {
  return function DummyParticleNetwork() {
    return <div data-testid="particle-network">Particle Network Mock</div>;
  };
});

describe("SignInForm integration tests", () => {
  let store;
  let mock;
  let consoleErrorSpy;

  beforeEach(() => {
    // Spy on console.error and silence it
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    
    mock = new MockAdapter(axios);
    store = configureStore({
      reducer: {
        auth: authReducer,
      },
    });

    render(
      <Provider store={store}>
        <BrowserRouter>
          <SignInForm />
        </BrowserRouter>
      </Provider>
    );
  });

  afterEach(() => {
    mock.reset();
    // Restore console.error
    consoleErrorSpy.mockRestore();
  });

  it("renders all form elements correctly", () => {
    expect(screen.getByRole('heading', { name: "Sign In" })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: "Admin" })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: "Manager" })).toBeInTheDocument();
    expect(screen.getByTestId("email-input")).toBeInTheDocument();
    expect(screen.getByTestId("password-input")).toBeInTheDocument();
    expect(screen.getByTestId("sign-in-button")).toBeInTheDocument();
  });

  it("toggles password visibility", () => {
    const passwordInput = screen.getByTestId("password-input");
    const toggleButton = screen.getByTestId("toggle-password");

    expect(passwordInput.type).toBe("password");
    fireEvent.click(toggleButton);
    expect(passwordInput.type).toBe("text");
    fireEvent.click(toggleButton);
    expect(passwordInput.type).toBe("password");
  });

  it("switches between admin and manager roles", () => {
    const adminButton = screen.getByRole('button', { name: "Admin" });
    const managerButton = screen.getByRole('button', { name: "Manager" });

    expect(adminButton).toHaveClass("active");
    expect(managerButton).not.toHaveClass("active");

    fireEvent.click(managerButton);
    expect(adminButton).not.toHaveClass("active");
    expect(managerButton).toHaveClass("active");
  });

  it("shows validation error for empty fields", async () => {
    fireEvent.click(screen.getByTestId("sign-in-button"));
    expect(await screen.findByTestId("form-error")).toHaveTextContent(
      "Please enter both email and password."
    );
  });

  it("logs in successfully with correct credentials", async () => {
    const mockResponse = { 
      user: { email: "test@example.com" },
      token: "123" 
    };
    mock
      .onPost("https://hivve.westus.cloudapp.azure.com/api/v1/auth/login")
      .reply(200, mockResponse);

    fireEvent.click(screen.getByRole('button', { name: "Manager" }));
    fireEvent.change(screen.getByTestId("email-input"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByTestId("password-input"), {
      target: { value: "password" },
    });

    fireEvent.click(screen.getByTestId("sign-in-button"));

    await waitFor(() => {
      const state = store.getState().auth;
      expect(state.user).toEqual(mockResponse.user);
      expect(state.loading).toBe(false);
      expect(state.error).toBeNull();
    });
  });

  it("shows error on invalid login credentials", async () => {
    const errorMessage = "Invalid credentials";
    mock
      .onPost("https://hivve.westus.cloudapp.azure.com/api/v1/auth/login")
      .reply(401, { message: errorMessage });

    fireEvent.change(screen.getByTestId("email-input"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByTestId("password-input"), {
      target: { value: "wrongpassword" },
    });

    fireEvent.click(screen.getByTestId("sign-in-button"));

    await waitFor(() => {
      const state = store.getState().auth;
      expect(state.user).toBeNull();
      expect(state.error).toBe(errorMessage);
    });

    expect(await screen.findByTestId("server-error")).toHaveTextContent(
      errorMessage
    );
  });

  it("handles network error gracefully", async () => {
    mock
      .onPost("https://hivve.westus.cloudapp.azure.com/api/v1/auth/login")
      .networkError();

    fireEvent.change(screen.getByTestId("email-input"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByTestId("password-input"), {
      target: { value: "password" },
    });

    fireEvent.click(screen.getByTestId("sign-in-button"));

    await waitFor(() => {
      const state = store.getState().auth;
      expect(state.user).toBeNull();
      expect(state.error).toBe("Login failed");
    });

    expect(await screen.findByTestId("server-error")).toHaveTextContent(
      "Login failed"
    );
  });

  it("shows loading state during authentication", async () => {
    mock
      .onPost("https://hivve.westus.cloudapp.azure.com/api/v1/auth/login")
      .reply(() => new Promise((resolve) => setTimeout(() => resolve([200, {}]), 100)));

    fireEvent.change(screen.getByTestId("email-input"), {
      target: { value: "test@example.com" },
    });
    fireEvent.change(screen.getByTestId("password-input"), {
      target: { value: "password" },
    });

    fireEvent.click(screen.getByTestId("sign-in-button"));
    
    expect(screen.getByText("Signing In...")).toBeInTheDocument();
    expect(screen.getByTestId("sign-in-button")).toBeDisabled();
  });
});
