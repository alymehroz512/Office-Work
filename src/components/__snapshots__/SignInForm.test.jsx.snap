// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm integration tests renders correctly 1`] = `null`;
